integrationTests/chainSimulator/mempool/mempool_test.go

```diff
@@ -485,3 +486,874 @@ func TestMempoolWithChainSimulator_Eviction(t *testing.T) {
 }
+
+func Test_Selection_ShouldNotSelectSameTransactionsWithSameSender(t *testing.T) {
+	t.Parallel()
+
+	host := txcachemocks.NewMempoolHostMock()
+	txpool, err := txcache.NewTxCache(txcache.ConfigSourceMe{
+		Name:                        "test",
+		NumChunks:                   16,
+		NumBytesThreshold:           maxNumBytesUpperBound,
+		NumBytesPerSenderThreshold:  maxNumBytesPerSenderUpperBoundTest,
+		CountThreshold:              math.MaxUint32,
+		CountPerSenderThreshold:     math.MaxUint32,
+		EvictionEnabled:             false,
+		NumItemsToPreemptivelyEvict: 1,
+		TxCacheBoundsConfig: config.TxCacheBoundsConfig{
+			MaxNumBytesPerSenderUpperBound: maxNumBytesPerSenderUpperBoundTest,
+		},
+	}, host)
+
+	require.Nil(t, err)
+	require.NotNil(t, txpool)
+
+	// create the non-virtual selection session, assure we have enough balance
+	selectionSession := createMockSelectionSession(map[string]*accountInfo{
+		"alice": {
+			balance: oneEGLD,
+			nonce:   0,
+		},
+		"receiver": {
+			balance: big.NewInt(0),
+			nonce:   0,
+		},
+	})
+
+	maxNumTxs := 2
+	options := holders.NewTxSelectionOptions(
+		10_000_000_000,
+		maxNumTxs,
+		selectionLoopMaximumDuration,
+		10,
+	)
+
+	poolTxs := maxNumTxs * 2
+	txHashes := make([][]byte, 0, poolTxs)
+
+	nonceTracker := newNoncesTracker()
+	for i := 0; i < poolTxs; i++ {
+		tx := &transaction.Transaction{
+			Nonce:     nonceTracker.getThenIncrementNonceByStringAddress("alice"),
+			Value:     oneQuarterOfEGLD,
+			SndAddr:   []byte("alice"),
+			RcvAddr:   []byte("receiver"),
+			Data:      []byte{},
+			GasLimit:  50_000,
+			GasPrice:  1_000_000_000,
+			ChainID:   []byte(configs.ChainID),
+			Version:   2,
+			Signature: []byte("signature"),
+		}
+
+		txHash := []byte(fmt.Sprintf("txHash%d", i))
+		txpool.AddTx(&txcache.WrappedTransaction{
+			Tx:               tx,
+			TxHash:           txHash,
+			SenderShardID:    0,
+			ReceiverShardID:  0,
+			Size:             0,
+			Fee:              big.NewInt(int64(tx.GasLimit * tx.GasPrice)),
+			PricePerUnit:     0,
+			TransferredValue: tx.Value,
+			FeePayer:         tx.SndAddr,
+		})
+
+		txHashes = append(txHashes, txHash)
+	}
+
+	blockBody := block.Body{MiniBlocks: []*block.MiniBlock{
+		{
+			TxHashes: txHashes[:len(txHashes)/2],
+		},
+	}}
+
+	require.Equal(t, txpool.CountTx(), uint64(4))
+
+	// do the first selection, first two txs should be returned
+	selectedTransactions, _ := txpool.SelectTransactions(selectionSession, options, defaultBlockchainInfo)
+	require.Equal(t, 2, len(selectedTransactions))
+	require.Equal(t, "txHash0", string(selectedTransactions[0].TxHash))
+	require.Equal(t, "txHash1", string(selectedTransactions[1].TxHash))
+
+	// propose the block
+	err = txpool.OnProposedBlock([]byte("blockHash1"), &blockBody, &block.Header{
+		Nonce:    0,
+		PrevHash: []byte("blockHash0"),
+		RootHash: []byte(fmt.Sprintf("rootHash%d", 0)),
+	})
+	require.Nil(t, err)
+
+	blockchainInfo := holders.NewBlockchainInfo([]byte("blockHash0"), 1)
```

Comment:
Above, you've written some comments to tell the story. Maybe add a small comment above this section, as well? E.g. "do the second selection, ...".